src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProjectTracker.cs

```diff
@@ -16,22 +18,53 @@ internal sealed partial class VisualStudioProjectTracker
         private readonly VisualStudioProjectFactory _projectFactory;
         internal IThreadingContext ThreadingContext { get; }
 
-        internal ImmutableArray<AbstractProject> ImmutableProjects => ImmutableArray<AbstractProject>.Empty;
+        [Obsolete("This is a compatibility shim for Live Share; please do not use it.")]
+        internal ImmutableArray<AbstractProject> ImmutableProjects => _projects.Values.ToImmutableArray();
 
         internal HostWorkspaceServices WorkspaceServices => _workspace.Services;
 
-        [Obsolete("This is a compatibility shim for TypeScript; please do not use it.")]
-        private readonly Dictionary<ProjectId, AbstractProject> _typeScriptProjects = new Dictionary<ProjectId, AbstractProject>();
+        [Obsolete("This is a compatibility shim for TypeScript and Live Share; please do not use it.")]
+        private readonly Dictionary<ProjectId, AbstractProject> _projects = new Dictionary<ProjectId, AbstractProject>();
 
+        [Obsolete("This is a compatibility shim; please do not use it.")]
         public VisualStudioProjectTracker(Workspace workspace, VisualStudioProjectFactory projectFactory, IThreadingContext threadingContext)
         {
             _workspace = workspace;
             _projectFactory = projectFactory;
             ThreadingContext = threadingContext;
+            DocumentProvider = new DocumentProvider(this, null, null);
+        }
+
+        [Obsolete("This is a compatibility shim for Live Share; please do not use it.")]
+        public VisualStudioProjectTracker(IServiceProvider serviceProvider, Workspace workspace)
+        {
+            _workspace = workspace;
+            ThreadingContext = serviceProvider.GetMefService<IThreadingContext>();
+
+            // This is used by Live Share to target their own workspace which is not the standard VS workspace; as a result this shouldn't have a project factory
+            // at all, because that's only targeting the VisualStudioWorkspace.
+            _projectFactory = null;
+
+            // We don't set DocumentProvider, because Live Share creates their own and then sets it later.
         }
 
         [Obsolete("This is a compatibility shim for TypeScript; please do not use it.")]
-        public DocumentProvider DocumentProvider => new DocumentProvider();
+        public DocumentProvider DocumentProvider { get; set; }
+
+        public Workspace Workspace => _workspace;
```

Comment:
Do you need `[Obsolete]` here as well?

Reply:
We could put it. We could also just delete this in Preview 2. :smile:

We've already got bits out for testing now, and I'm not sure a respin is worth it, but good catch.